### ISSUE TITLE
Required to always use --debug-brk for node debugging?

### ISSUE THREAD

**Issue by andrewconnell** [Microsoft/vscode-node-debug#78]
TLDR: 
Is it a set requirement that to debug Node.js applications you must start them (either in when *launch* or *attach*) with `--debug-brk`? It seems this is required... the behavior is almost like when the breakpoint is set in a `*.ts`, VSCode maps it to the equiv. `*.js` using the sourcemap... but if it doens't have the sourcemap, it doesn't resolve it... as in the case when you create a breakpoint BEFORE launching / attaching to a running process.

Result - you can't keep breakpoints between sessions... you have to clear them all out each time OR toggle them on/off when starting a new debugging session.

MORE:
The only way I can hit a breakpoint in a TypeScript file is if I do the following:
1. start the node process using `--debug-brk`... once VSCode hits the first line in the process...
1. I then set the breakpoint(s)
1. Hit F5 to go to the first breakpoint

If I set the breakpoint BEFORE attaching, VSCode won't hit the breakpoint (grey breakpoint because it can't find the sourcemap... but the sourcemap is valid... proven by using the MORE method above with no change to the source `*.ts` or transpiled `*.js` files.

If I set the breakpoint AFTER attaching, it works exactly how you'd expect it.

> EDIT: BTW, including / excluding `--nolazy` has zero effect

Using Node.js v5.11.1, VSCode 1.2.1 & testing this using a simple console app that writes a message to the console.

**Comment by weinand** [comment]
@andrewconnell the `--debug-brk` option is only useful if you have breakpoints in the startup phase of your program. With `--debug-brk` node.js stops at the entry point of your program and a debugger has enough time to register breakpoints for code locations that are immediately following the entry point.

If node.js is started with the `--debug` it does **not** stop execution at the entry point of your program and the debugger might come too late in registering breakpoints for early code locations.

Above you said: "you can't keep breakpoints between sessions... you have to clear them all out each time OR toggle them on/off when starting a new debugging session."

This is only true if something is broken in your setup or you are using dynamically created sourcemaps as those used by SystemJS or babel register.

If VS Code says that it cannot find the sourcemap (and the breakpoints stays grey), then the setup is wrong. Please make sure that the `outDir` attribute points to the root folder of the generated code.

**Comment by andrewconnell** [comment]
@weinand Sorry for the delay in responding... was at a conference. At any rate... thanks for the response. I've added the `outDir` attribute that points where the generated code is. For my config, sourcemaps aren't dynamically built. Like most TS projects, I transpile with gulp, but I have a programmatic build that doesn't use `tsconfig.json` exactly... I modify the TypeScript project in the build to add additional excludes & the destination folder for output.

Even with the tweak to the `tsconfig.json` for the `outDir`, I'm still having the same experience.

VSCode never sees my sourcemaps, even when the sourcemap is there UNLESS I have first have VSCdoe hit a breakpoint within the JS generated files, and THEN I can set (or toggle on/off) the TypeScript breakpoints for VSCode to see them. I can do this by either setting `--debug-brk` or since it's a web app, put it in a route that I'm working on.

Working on a video (as I admit my build is a bit unconventional, but I don't think its impacting the debugging aspect) and project I can let you experiment on... will follow up for repo.

**Comment by andrewconnell** [comment]
Hi @weinand ... hopefully this will help you see what I'm trying to say... ideally you see something I'm doing wrong and it's not a bug in VSCode, but if so, happy to help. Here's a short video explaining what I'm seeing:
https://youtu.be/FSXiIqhAjsw

Here's the repo where you can dupe it:
https://github.com/andrewconnell/vscode-node-debug-issue78

**Comment by weinand** [comment]
@andrewconnell thanks a lot for making your project available and for the screencast.

After setting `outDir` to `"${workspaceRoot}/dist"` breakpoints were working just fine for me in VS Code 1.2.1: you can easily see the difference: start debugging with F5 and then set a breakpoint in homeController.ts:38. Without `outDir` set, the breakpoints become grey (unverified). With `outDir` set they stay red (verified).

In addition you can set `program` to the typescript file `"${workspaceRoot}/src/server/index.ts" instead of the generated JavaScript file in dist.

And I noticed that the `sourcesContent` array in the inlined sourcemaps has a bogus `null` entry:
```json
 "sourcesContent": [null]
```
This probably doesn't affect VS Code but it is still not following the spec.

**Comment by andrewconnell** [comment]
Dang... thanks @weinand ... I wasn't aware the `outDir` needed to be set in the `launch.json`. I thought that was being fed from `tsconfig.json`. The token in the string replacement in your response was the clue. You're right... I have it working now. Thanks! Going to go back and look at the docs... possibly submitting a PR to clarify the docs.

Wasn't aware the `program` property could point to the source TypeScript file. Thanks!

Regarding the null inline TypeScript... thanks for mentioning it. Recent update to the gulp task I'm using must have screwed that up. They were being generated until recently ... I'll get that fixed. Thanks!
